testdata/p4_16_samples/parser_subparser_v1model.p4

```diff
@@ -0,0 +1,190 @@
+#include <v1model.p4>
+
+struct metadata { }
+
+header data_t {
+    bit<8> f;
+}
+
+header data_t16 {
+    bit<16> f;
+}
+
+struct headers {
+    data_t   h1;
+    data_t16 h2;
+    data_t   h3;
+    data_t   h4;
+}
+
+struct headers2 {
+    data_t h1;
+}
+
+parser PacketParser(      packet_in packet,
+                    out   headers   hdr,
+                    inout headers2  inout_hdr) {
+    headers2 local_hdr2;
+
+    state start {
+        packet.extract(hdr.h1);
+        transition select(hdr.h1.f) {
+            10: local_10;
+            11: local_11;
+            12: local_12;
+            default: accept;
+        }
+    }
+
+    state local_10 {
+        local_hdr2.h1.f = 10;
+        local_hdr2.h1.setValid();
+        transition parse_h2;
+    }
+
+    state local_11 {
+        local_hdr2.h1.f = 11;
+        local_hdr2.h1.setValid();
+        transition parse_h2;
+    }
+
+    state local_12 {
+        local_hdr2.h1.f = 12;
+        local_hdr2.h1.setValid();
+        transition parse_h2;
+    }
+
+    state parse_h2 {
+        inout_hdr.h1.f = local_hdr2.h1.f;
+        inout_hdr.h1.setValid();
+        packet.extract(hdr.h2);
+        transition accept;
+    }
+}
+
+parser ParserImpl(      packet_in           packet,
+                  out   headers             hdr,
+                  inout metadata            meta,
+                  inout standard_metadata_t standard_metadata) {
+    PacketParser() p;
+    headers2 hdr2;
+
+    state start {
+        transition select(standard_metadata.ingress_port) {
+            0: p_0;
+            1: p_1;
+            2: p_2;
+            3: p_3;
+            4: p_4;
+            5: p_5;
+            default: accept;
+        }
+    }
+
+    state p_0 {
+        hdr2.h1.f = 100;
+        hdr2.h1.setValid();
+
+        p.apply(packet, hdr, hdr2);
```

Comment:
in your example you always apply the sub-parser to the same arguments, which are toplevel variables in the caller.
Can you make a few simpler tests?
And some which use locals with the same name or different arguments?

Reply:
Ok, I'll work on more test cases.